docs/plugins/new-backend-system.md

```diff
@@ -0,0 +1,305 @@
+        catalog.addProcessor(new MyCustomProcessor());
+      },
+    });
+  },
+});
+```
+
+### Extension Points
+
+Modules depend on extension points just as a regular dependency but specifying it in the `deps` section.
+
+#### Defining an Extension Point
+
+```ts
+import { createExtensionPoint } from '@backstage/backend-plugin-api';
+
+export interface ScaffolderActionsExtensionPoint {
+  addAction(action: ScaffolderAction): void;
+}
+
+export const scaffolderActionsExtensionPoint =
+  createExtensionPoint<ScaffolderActionsExtensionPoint>({
+    id: 'scaffolder.actions',
+  });
+```
+
+#### Registering an Extension Point
+
+Extension points are registered by a plugin and extended by modules.
+
+## Backend Services
+
+The default backend provides several _services_ out of the box which includes access to configuration, logging, databases and more.
+Services are declared using their _serviceRef_ in the `deps` section of plugin or module requiring them and are then available in the `init` method of the plugin or module.
+
+### Service References
+
+A `ServiceRef` is a named reference to an interface which are later used to resolve the concrete service implementation. Conceptually this is very similar to `ApiRef`s in the frontend.
+Services is what provides common utilities that previously resided in the `PluginEnvironment` such as Config, Logging and Database.
+
+On startup the backend will make sure that the services are initialized before being passed to the plugin/module that depend on them.
+ServiceRefs contain a scope which is used to determine if the serviceFactory creating the service will create a new instance scoped per plugin/module or if it will be shared. `plugin` scoped services will be created once per plugin/module and `root` scoped services will be created once per backend instance.
+
+#### Defining a Service
+
+```ts
+import {
+  createServiceFactory,
+  pluginMetadataServiceRef,
+  loggerServiceRef,
+} from '@backstage/backend-plugin-api';
+import { ExampleImpl } from './ExampleImpl';
+
+export interface ExampleApi {
+  doSomething(): Promise<void>;
+}
+
+export const exampleServiceRef = createServiceRef<ExampleApi>({
+  id: 'example',
+  scope: 'plugin', // can be 'root' or 'plugin'
+
+  // The defaultFactory is optional to implement but it will be used if no other factory is provided to the backend.
+  // This is allows for the backend to provide a default implementation of the service without having to wire it beforehand.
+  defaultFactory: async service =>
+    createServiceFactory({
+      service,
+      deps: {
+        logger: loggerServiceRef,
+        plugin: pluginMetadataServiceRef,
+      },
+      // Logger is available directly in the factory as it's a root scoped service and will be created once per backend instance.
+      async factory({ logger }) {
+        // plugin is available as it's a plugin scoped service and will be created once per plugin.
+        return async ({ plugin }) => {
```

Comment:
This is a  big nitpick since this is just example code but linting will probably complain here that `plugin` is never used 😅 maybe pass it into `ExampleImpl` as well?